llvm/unittests/ADT/StringRefTest.cpp

```diff
@@ -16,20 +16,7 @@
 #include "gtest/gtest.h"
 using namespace llvm;
 
-namespace llvm {
-
-std::ostream &operator<<(std::ostream &OS, const StringRef &S) {
```

Comment:
These are not unused.
They are used by gtest on a test failure to print human readable representation of StringRefs.
They must be in llvm namespace.

They *should* be moved to a place where *all* GTests can see them, otherwise there is an ODR violation as explained [here](https://stackoverflow.com/questions/24673515/googletest-printto-not-getting-called-for-a-class) and they may fail to work.

Reply:
Thanks. I guess it this part:

>This has caught me out, so I'm really glad to find this solution. My case was that MyType has an ostream& operator<<(), but it's defined in mytype_io.h rather than mytype.h. If any unit test cpp file uses MyType but doesn't include mytype_io.h then it will cause the default template implementation in gtest-printers.h to be instantiated.

>So my type wasn't being printed correctly. The fix was to ensure that all unit tests that used MyType included mytype_io.h.

So without the << operators, it will call some default printers which won't print the StringRef the right way. May be I can add a comment to that effect?

Reply:
> So without the << operators, it will call some default printers which won't print the StringRef the right way.

They may or may not work, depending on which instantiation is picked by the linker.

> May be I can add a comment to that effect?

A comment would be great, thanks.

Reply:
Hmm, I can see that even without these, the StringRef and the pair gets printed properly:

```
[ RUN      ] StringRefTest.Construction
/home/rjoshi/upstream_llvm/llvm-project/llvm/unittests/ADT/StringRefTest.cpp:41: Failure
Expected equality of these values:
  ""
    Which is: 0x55db55e939d4
  StringRef("a")
    Which is: "a"
```

and

```
/home/rjoshi/upstream_llvm/llvm-project/llvm/unittests/ADT/StringRefTest.cpp:166: Failure
Expected equality of these values:
  std::make_pair(StringRef("he"), StringRef("lo"))
    Which is: ("he", "lo")
  Str.split('y')
    Which is: ("hello", "")
```

Reply:
SG, I'll restore. In any case its an overreach for what the change claims to do.

Reply:
Thanks for checking. This is probably because there is another printer [here](https://github.com/llvm/llvm-project/blob/a866ce789eb99da4d7a486eeb60a53be6c75f4fd/third-party/unittest/googletest/include/gtest/internal/custom/gtest-printers.h#L52) and it takes precedence.

I wasn't aware of this header until now. I think the overloaded << for StringRef can safely be removed then.
(and I guess `pair<StringRef, StringRef>` overload is still needed)

Reply:
Doh! I'll remove them again :) What about pair<>? Is there a generic pair printer in the test framework?

Reply:
may be its [this ](https://github.com/google/googletest/blob/main/googletest/include/gtest/gtest-printers.h#L812)for pair.

Reply:
Restored deleting of << operators.

Reply:
Updated change description as well to say:

* Delete unused << operators (googltest framework provides printers for StringRef and pairs).